Structural Refinement and Config Update

ange-black69 opened the pull request:
- Using a proper singleton design pattern !
- Prevent player command when joining an arena (you can do /ba, /arena /blockyarena even tho)
- Added a message to broadcast others player when Someone joining a arena channel
- Added a simple confg
- Configurable Lobby Timeout
- Implemented a fully working reload listener 
 -> if you change a node in the config and do sponge plugins reload it update node in memory !

KNOW BUG : 
- No persistant config : If you change a config node and you restart the server, the value is overrided (maybe check if file exists before writting a new empty node ? using loader.load() I dont understand this crap lol )

ange-black69 committed:
Prevent irrelevant commands while playing

Players cannot execute irrelevant commands while in an active game session. Lobby timeout can be configured with the default config.

Hide internal impl of ConfigManager

ConfigManager should provide an abstraction for the rest of the project on accessing configurations instead of spreading its internal rootNode all over the project. The rest of the project should be alleviated from remembering a set of train wreck method calls just to get a node from the config.

mrdarcychen commented on src/main/java/net/huskycraft/blockyarena/games/Game.java:
```diff
@@ -144,12 +143,17 @@ private void inspect() {
+            
         }
+        else {
+        	MessageChannel.TO_ALL.send((Text)Text.builder("New arena game ! join with /ba join solo !").color(TextColors.GOLD).build());
```
I love the idea of broadcasting new available game session to other players. However, we might need to deal with 2v2 so the message needed be adjusted accordingly (`/ba join doubles`).

We might also consider giving players the option to tune out this announcement. Don't send this message to the player that just executed the join command.

mrdarcychen commented on src/main/java/net/huskycraft/blockyarena/listeners/EntityListener.java:
```diff
@@ -37,19 +37,60 @@
+		if (game != null)
+		{
+			
+				if (game.getGameState() == GameState.RECRUITING || game.getGameState() == GameState.STARTED || game.getGameState() == GameState.STARTING || game.getGameState() == GameState.STOPPING) 
```
```suggestion
				if (gamer.getStatus() == GamerStatus.PLAYING)
```

ange-black69 replied:
Fixed !

mrdarcychen commented on src/main/java/net/huskycraft/blockyarena/listeners/EntityListener.java:
```diff
@@ -37,19 +37,60 @@
+				if (game.getGameState() == GameState.RECRUITING || game.getGameState() == GameState.STARTED || game.getGameState() == GameState.STARTING || game.getGameState() == GameState.STOPPING) 
+				{
+						
+					MessageChannel.TO_ALL.send((Text)Text.builder("command issued : " + command).color(TextColors.GOLD).build());
```
Do not broadcast commands issued by individual players. May leak sensitive information to others. If this is for testing purposes, remember to mark it as TODO and remove it afterwards.

ange-black69 replied:
Removed !
Was a test until I properly debug !

mrdarcychen commented on .settings/org.eclipse.buildship.core.prefs:
```diff
@@ -0,0 +1,2 @@
+connection.project.dir=
```
add this file to .gitignore

mrdarcychen commented on .classpath:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
add this file to .gitignore

mrdarcychen commented on .project:
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
add this file to .gitignore

mrdarcychen commented on src/main/java/net/huskycraft/blockyarena/BlockyArena.java:
```diff
@@ -122,15 +148,48 @@ private void createDirectories() {
         }
     }
+    
+    private void HandleConfiguration()
```
Consider moving the following to a standalone class (a class dedicated for configuration loading). Right now the main class is overwhelmed.

ange-black69 replied:
Done : 
-> ConfigManager Singleton Class
-> create config file if not exist and fill default node value
-> load the conf file if it exist !

mrdarcychen commented on src/main/java/net/huskycraft/blockyarena/BlockyArena.java:
```diff
@@ -122,15 +148,48 @@ private void createDirectories() {
+    	
+    	loader = HoconConfigurationLoader.builder().setPath(defaultConfig).build();
+    	
+         this.rootNode = loader.createEmptyNode(ConfigurationOptions.defaults());
```
Inconsistent indentation

mrdarcychen commented on src/main/java/net/huskycraft/blockyarena/BlockyArena.java:
```diff
@@ -122,15 +148,48 @@ private void createDirectories() {
+    	loader = HoconConfigurationLoader.builder().setPath(defaultConfig).build();
+    	
+         this.rootNode = loader.createEmptyNode(ConfigurationOptions.defaults());
+    	 rootNode.getNode("timers", "lobby", "cooldownSec").setValue(15);
```
Everytime the server inits, this code gets called and thus override the existing value in config to 15. Before setting the default to 15, you may want to check if default config already exists.

ange-black69 replied:
Fixed as of new configuration manager !

mrdarcychen commented on src/main/java/net/huskycraft/blockyarena/BlockyArena.java:
```diff
@@ -109,6 +134,7 @@ private void createManagers() {
     private void createDirectories() {
         arenaDir = Paths.get(getConfigDir().toString() + "/arenas");
         kitDir = Paths.get(getConfigDir().toString() + "/kits");
+        defaultConfig = Paths.get(getConfigDir().toString(), "default.json");
```
To be consistent with Sponge's style, we're using .conf for configs.

ange-black69 replied:
Fixed

mrdarcychen reviewed (changes requested):
Thank you for initiating this pull request. Most added features work as expected. Please note the following for your revision in addition to the in-line comments in this pull request:
- Exclude configuration files (dot files) that are only relevant to your local development environment.
- We do not have automated tests to enforce coding style yet, but for now please follow this [guide](https://docs.spongepowered.org/stable/en/contributing/implementation/codestyle.html).]

We'll work together and establish some community guidelines for working on this project later down the road. Cheers!

Commit:
Update copyright notice and source code headers

Replace MIT with Apache License 2.0. To help well-intentioned people avoid accidental copying that doesn’t comply with the license, proper license headers are added to all source files.

Commit:
Deflate Game class with state pattern

This is the first attempt at refactoring the Game class. Gameflow follows this order: Entering -> Starting -> Playing -> Stopping -> Leaving. Game class now acts as the context for each of the five states.

ange-black69 committed:
Decouple setup logic from main class

Deprecate enums in favor of the new State pattern. Added server
listener, optimize imports, and create standalone managers for certain
tasks.

Commit:
Add preliminary support for FFA mode

An arena can have more than two start points.

Commit:
Properly dismiss players on LeavingState

Sometimes players were not be able to exit properly and stuck in spectator mode even after the game ended. This was due to a concurrent modification of the gamers list, and has now been resolved.
